base_layer/wallet/src/ffi.rs

```diff
@@ -29,87 +29,402 @@
+        Box::from_raw(pk);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn public_key_get_key(pk: *mut WalletPublicKey) -> *mut ByteVector {
+    let mut bytes = ByteVector(Vec::new());
+    if !pk.is_null() {
+        bytes.0 = (*pk).to_vec();
+    }
+    Box::into_raw(Box::new(bytes))
+}
+
+/// -------------------------------------------------------------------------------------------- ///
+
+/// -------------------------------- Private Key ----------------------------------------------- ///
+pub type WalletPublicKey = PrivateKey;
+
+#[no_mangle]
+pub unsafe extern "C" fn privatekey_create(bytes: *mut ByteVector) -> *mut WalletPrivateKey {
+    let mut v = Vec::new();
+    if !bytes.is_null() {
+        v = (*bytes).0.clone();
+    }
+    let pk = WalletPrivateKey::from_bytes(&v).unwrap();
+    Box::into_raw(Box::new(pk))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn privatekey_destroy(pk: *mut WalletPrivateKey) {
+    if !pk.is_null() {
+        Box::from_raw(pk);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn privatekey_get_key(pk: *mut WalletPrivateKey) -> *mut ByteVector {
+    let mut bytes = ByteVector(Vec::new());
+    if !pk.is_null() {
+        bytes.0 = (*pk).to_vec();
+    }
+    Box::into_raw(Box::new(bytes))
+}
+
+/// -------------------------------------------------------------------------------------------- ///
+
+/// -------------------------------------- OutputManagerConfig --------------------------------- ///
+pub type WalletOutputManagerConfig = OutputManagerConfig;
+
+#[no_mangle]
+pub unsafe extern "C" fn outputmanagerconfig_create(
+    key: *mut PrivateKey,
+    b_seed: *mut c_char,
+    pki: c_ulonglong,
+) -> *mut WalletOutputManagerConfig
+{
+    let mut rng = rand::OsRng::new().unwrap();
+    let mut k = PrivateKey::random(&mut rng);
+
+    if !key.is_null() {
+        k = (*key).clone();
+    }
+
+    let mut str = CString::new("").unwrap().to_str().unwrap().to_owned();
+    if !b_seed.is_null() {
+        str = CStr::from_ptr(b_seed).to_str().unwrap().to_owned();
+    }
+
+    let omc = WalletOutputManagerConfig {
+        master_key: k,
+        branch_seed: str.to_string(),
+        primary_key_index: pki as usize,
+    };
+    Box::into_raw(Box::new(omc))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn outputmanagerconfig_destroy(wc: *mut WalletOutputManagerConfig) {
+    if !wc.is_null() {
+        Box::from_raw(wc);
+    }
+}
+/// ---------------------------------------------------------------------------------------------///
+
+/// ----------------------------------- PeerFeature -------------------------------------------- ///
+pub type WalletPeerFeatures = PeerFeatures;
+
+#[no_mangle]
+pub unsafe extern "C" fn peerfeatures_create() -> *mut WalletPeerFeatures {
+    let pf = WalletPeerFeatures::new(Vec::new());
+    Box::into_raw(Box::new(pf))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn peerfeatures_add(pf: *mut WalletPeerFeatures, feature: c_uchar) {
+    if !pf.is_null() {
+        match feature {
```

Comment:
We will need to come up with a better solution than this because if this Enum changes then this breaks. We will do this by writing from() implementations to convert these structs into strings and back to rust objects that don't rely on the manual process done here.